query/src/pipelines/processors/processor_dag.rs

```diff
@@ -0,0 +1,369 @@
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_planners::AggregatorFinalPlan;
+use common_planners::AggregatorPartialPlan;
+use common_planners::ExpressionPlan;
+use common_planners::FilterPlan;
+use common_planners::HavingPlan;
+use common_planners::LimitByPlan;
+use common_planners::LimitPlan;
+use common_planners::PlanNode;
+use common_planners::ProjectionPlan;
+use common_planners::ReadDataSourcePlan;
+use common_planners::RemotePlan;
+use common_planners::SelectPlan;
+use common_planners::SortPlan;
+use petgraph::dot::Config;
+use petgraph::dot::Dot;
+use petgraph::prelude::NodeIndex;
+use petgraph::prelude::StableGraph;
+
+use crate::api::FlightTicket;
+use crate::pipelines::processors::processor::ProcessorRef;
+use crate::pipelines::processors::Processor;
+use crate::pipelines::transforms::AggregatorFinalTransform;
+use crate::pipelines::transforms::AggregatorPartialTransform;
+use crate::pipelines::transforms::ExpressionTransform;
+use crate::pipelines::transforms::FilterTransform;
+use crate::pipelines::transforms::GroupByFinalTransform;
+use crate::pipelines::transforms::GroupByPartialTransform;
+use crate::pipelines::transforms::LimitByTransform;
+use crate::pipelines::transforms::LimitTransform;
+use crate::pipelines::transforms::ProjectionTransform;
+use crate::pipelines::transforms::RemoteTransform;
+use crate::pipelines::transforms::SortMergeTransform;
+use crate::pipelines::transforms::SortPartialTransform;
+use crate::pipelines::transforms::SourceTransform;
+use crate::sessions::DatabendQueryContextRef;
+
+pub struct ProcessorsDAG {
+    graph: StableGraph<Arc<dyn Processor>, ()>,
+}
+
+impl ProcessorsDAG {
+    pub fn create(graph: StableGraph<Arc<dyn Processor>, ()>) -> ProcessorsDAG {
+        ProcessorsDAG { graph }
+    }
+}
+
+impl Debug for ProcessorsDAG {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{:?}",
+            Dot::with_config(&self.graph, &[Config::EdgeNoLabel])
+        )
+    }
+}
+
+// TODO(Winter): import distributed processors in DAG. We can use edge to describe them. e.g. enum EdgeAttrs { Local, Remote }
```

Comment:
Maybe this information should be encapsulated in `Processor` nodes, so we don't need to be aware of edge labels?

Reply:
We may also need `ProcessorsDAGExecutor`. It may should control the data flow to using the properties of the graph(maybe not the properties of specific type graph node?). welcome comment : )

Reply:
> `ProcessorsDAGExecutor`

What's the difference between current executor and `ProcessorDAGExecutor`?

By the way, I'm not quite sure about the motivation of using `petgraph` to describe `Processor`. Would you mind to elaborate that for me?

Reply:
Is this patch intended to impl `EXPLAIN PIPELINE graph = 1` and show the graph format output?